packages/utils/other-utils/src/set-path-utils.js

```diff
@@ -0,0 +1,139 @@
+ * @param {array} cellSetSelection An array of cell set paths that are selected.
+ * @returns An array of paths that should be displayed.
+ */
+export function generateCellSetPaths(
+  mergedCellSets,
+  hierarchy,
+  cellSetExpansion,
+  cellSetSelection,
+) {
+  const paths = getPaths({ children: mergedCellSets.tree });
+
+  // returns true if path is contained in allPaths, false otherwise
+  const contains = (allPaths, path) => allPaths.some(p => p.toString() === path.toString());
+
+  return paths.filter((clusterPath) => {
+    // clusterPath is a parent of some selected cell set and is expanded. We should discard it.
+    if (contains(cellSetExpansion, clusterPath)) return false;
+
+    // clusterPath is not selected. Now we need to determine if we should keep it.
+    if (!contains(cellSetSelection, clusterPath)) {
+      /* This line checks if the longest super set of clusterPath in cellSetSelection is longer
+            than clusterPath itself. If true, it means clusterPath is a parent of some selected cell
+            set but not expanded and we should discard it. */
+      if (findLongest(cellSetSelection, clusterPath, false).length > clusterPath.length) {
+        return false;
+      }
+
+      // the clusterPath is too deep in the tree. We should discard it.
+      if (cellSetExpansion.length === 0 && clusterPath.length > 2) return false;
+
+      const longestSubset = findLongest(cellSetExpansion, clusterPath, true);
+      // another case of the clusterPath being deep in the tree. We should discard it.
+      if (cellSetExpansion.length > 0 && longestSubset.length + 1 < clusterPath.length) {
+        return false;
+      }
+    }
+    return clusterPath[0] === hierarchy[0];
+  });
+}
+
+/**
+ * Finds the hierarchy, where cell sets were most recently selected/deselected.
+ * @param {array} lastSelection All selected cell sets prior to the modification.
+ * @param {array} currentSelection All cell sets that are currently selected.
+ * @returns The name of the hierarchy, within which the most recent change occured.
+ * If nothing changed, returns 0.
+ */
+export function findChangedHierarchy(prevSelectedPaths, currSelectedPaths) {
+  const arrOfPathsToStr = subarray => subarray.toString();
+
+  const strOfPathsToArr = subarray => subarray.split(',').map((element) => {
+    const num = Number(element);
+    return num === parseFloat(element) ? num : element;
+  });
+
+  const prevPathsStrings = prevSelectedPaths.map(arrOfPathsToStr);
+  const currPathsStrings = currSelectedPaths.map(arrOfPathsToStr);
+
+  const unselectedPathsStr = prevPathsStrings.filter(
+    pathAsStr => !currPathsStrings.includes(pathAsStr),
+  );
+  const newlySelectedPathsStr = currPathsStrings.filter(
+    pathAsStr => !prevPathsStrings.includes(pathAsStr),
+  );
+
+  // nothing was selected or unselected, therefore hierarchy stays the same
+  if (unselectedPathsStr.length === 0 && newlySelectedPathsStr.length === 0) {
+    return 0;
```

Comment:
```suggestion
    return null;
```

Goes along with other comment about false-y values